<?xml version="1.0" encoding="ISO-8859-1"?>

<bugrepository name="canal">
	<bug id="330" opendate="2017-06-22 03:57:23" fixdate="2017-08-03 13:15:38">
		<buginformation>
			<summary>MySQL5.7 JSON解析问题</summary>
			<description>canal 版本1.0.24 写入MySQL的JSON数据，Canal解析失败  2017-06-22 11:39:25.104 [destination = example , address = /127.0.0.1:3306 , EventParser] ERROR com.alibaba.otter.canal.common.alarm.LogAlarmHandler - destination:example[com.alibaba.otter.canal.parse.exception.CanalParseException: com.alibaba.otter.canal.parse.exception.CanalParseException: parse row data failed. Caused by: com.alibaba.otter.canal.parse.exception.CanalParseException: parse row data failed. Caused by: java.lang.IllegalArgumentException: illegal json data         at com.taobao.tddl.dbsync.binlog.JsonConversion.parse_array_or_object(JsonConversion.java:81)         at com.taobao.tddl.dbsync.binlog.JsonConversion.parse_value(JsonConversion.java:61)         at com.taobao.tddl.dbsync.binlog.event.RowsLogBuffer.fetchValue(RowsLogBuffer.java:955)         at com.taobao.tddl.dbsync.binlog.event.RowsLogBuffer.nextValue(RowsLogBuffer.java:99)         at com.alibaba.otter.canal.parse.inbound.mysql.dbsync.LogEventConvert.parseOneRow(LogEventConvert.java:495)         at com.alibaba.otter.canal.parse.inbound.mysql.dbsync.LogEventConvert.parseRowsEvent(LogEventConvert.java:376)         at com.alibaba.otter.canal.parse.inbound.mysql.dbsync.LogEventConvert.parse(LogEventConvert.java:108)         at com.alibaba.otter.canal.parse.inbound.mysql.dbsync.LogEventConvert.parse(LogEventConvert.java:62)         at com.alibaba.otter.canal.parse.inbound.AbstractEventParser.parseAndProfilingIfNecessary(AbstractEventParser.java:326)         at com.alibaba.otter.canal.parse.inbound.AbstractEventParser$3$1.sink(AbstractEventParser.java:176)         at com.alibaba.otter.canal.parse.inbound.mysql.MysqlConnection.dump(MysqlConnection.java:130)         at com.alibaba.otter.canal.parse.inbound.AbstractEventParser$3.run(AbstractEventParser.java:209)         at java.lang.Thread.run(Thread.java:745)   另外如果json中包含双引号，canal解析之后丢失了转义符号。 ![image](https://user-images.githubusercontent.com/5357638/27416733-e8f73ba8-5741-11e7-96c0-2a0de5ddc8af.png) </description>
		</buginformation>
		<fixedfiles>
			<file>dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/JsonConversion.java</file>
			<file>dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java</file>
			<file>parse/src/main/java/com/alibaba/otter/canal/parse/inbound/AbstractEventParser.java</file>
		</fixedfiles>
	</bug>
	<bug id="274" opendate="2017-03-03 09:39:48" fixdate="2017-04-01 13:40:14">
		<buginformation>
			<summary>解析binlog日志失败</summary>
			<description>           canal运行后解析binlog日志的时候出现错误，提示Read Q_SQL_MODE_CODE error: limit excceed: 67 我看了canal的部分源码，貌似是sql_mode设置的不对？SQL_mode在canal里定义是64位的所以用getLong64，但是实际是67导致越界了？没看出数据库的数据有啥问题 PS: mysql版本是5.7.14   mysqlbinlog的日志： ------------------------------------------------------------------- 见附件 ![binlog](https://cloud.githubusercontent.com/assets/2059502/23546331/104ef6f0-003a-11e7-8f0c-40f50a85356c.png)  [binlog.txt](https://github.com/alibaba/canal/files/816403/binlog.txt)   canal实例日志： ----------------------------------------------------------------- 2017-03-03 17:17:19.952 [destination = gene_cachenotice_play3 , address = /172.16.10.213:3306 , EventParser] WARN  c.a.otter.canal.parse.inbound.mysql.MysqlEventParser - prepare to find start position just last position  {"identity":{"slaveId":-1,"sourceAddress":{"address":"172.16.10.213","port":3306}},"postion":{"included":false,"journalName":"mysql-bin.000016","position":24856433,"serverId":213,"timestamp":1488484784000}} 2017-03-03 17:17:19.959 [destination = gene_cachenotice_play3 , address = /172.16.10.213:3306 , EventParser] WARN  com.taobao.tddl.dbsync.binlog.LogDecoder - Decoding Query failed from: mysql-bin.000016:24858343 java.io.IOException: Read Q_SQL_MODE_CODE error: limit excceed: 67 at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.unpackVariables(QueryLogEvent.java:650) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.&lt;init&gt;(QueryLogEvent.java:477) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.LogDecoder.decode(LogDecoder.java:154) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.LogDecoder.decode(LogDecoder.java:106) ~[canal.parse.dbsync-1.0.23.jar:na] at com.alibaba.otter.canal.parse.inbound.mysql.MysqlConnection.dump(MysqlConnection.java:123) [canal.parse-1.0.23.jar:na] at com.alibaba.otter.canal.parse.inbound.AbstractEventParser$3.run(AbstractEventParser.java:209) [canal.parse-1.0.23.jar:na] at java.lang.Thread.run(Thread.java:745) [na:1.8.0_25] Caused by: java.lang.IllegalArgumentException: limit excceed: 67 at com.taobao.tddl.dbsync.binlog.LogBuffer.getLong64(LogBuffer.java:873) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.unpackVariables(QueryLogEvent.java:566) ~[canal.parse.dbsync-1.0.23.jar:na] ... 6 common frames omitted 2017-03-03 17:17:19.960 [destination = gene_cachenotice_play3 , address = /172.16.10.213:3306 , EventParser] ERROR c.a.otter.canal.parse.inbound.mysql.MysqlEventParser - dump address 172.16.10.213/172.16.10.213:3306 has an error, retrying. caused by  java.io.IOException: Read Q_SQL_MODE_CODE error: limit excceed: 67 at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.unpackVariables(QueryLogEvent.java:650) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.&lt;init&gt;(QueryLogEvent.java:477) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.LogDecoder.decode(LogDecoder.java:154) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.LogDecoder.decode(LogDecoder.java:106) ~[canal.parse.dbsync-1.0.23.jar:na] at com.alibaba.otter.canal.parse.inbound.mysql.MysqlConnection.dump(MysqlConnection.java:123) ~[canal.parse-1.0.23.jar:na] at com.alibaba.otter.canal.parse.inbound.AbstractEventParser$3.run(AbstractEventParser.java:209) ~[canal.parse-1.0.23.jar:na] at java.lang.Thread.run(Thread.java:745) [na:1.8.0_25] Caused by: java.lang.IllegalArgumentException: limit excceed: 67 at com.taobao.tddl.dbsync.binlog.LogBuffer.getLong64(LogBuffer.java:873) ~[canal.parse.dbsync-1.0.23.jar:na] at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.unpackVariables(QueryLogEvent.java:566) ~[canal.parse.dbsync-1.0.23.jar:na] ... 6 common frames omitted 2017-03-03 17:17:19.960 [destination = gene_cachenotice_play3 , address = /172.16.10.213:3306 , EventParser] ERROR com.alibaba.otter.canal.common.alarm.LogAlarmHandler - destination:gene_cachenotice_play3[java.io.IOException: Read Q_SQL_MODE_CODE error: limit excceed: 67 at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.unpackVariables(QueryLogEvent.java:650) at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.&lt;init&gt;(QueryLogEvent.java:477) at com.taobao.tddl.dbsync.binlog.LogDecoder.decode(LogDecoder.java:154) at com.taobao.tddl.dbsync.binlog.LogDecoder.decode(LogDecoder.java:106) at com.alibaba.otter.canal.parse.inbound.mysql.MysqlConnection.dump(MysqlConnection.java:123) at com.alibaba.otter.canal.parse.inbound.AbstractEventParser$3.run(AbstractEventParser.java:209) at java.lang.Thread.run(Thread.java:745) Caused by: java.lang.IllegalArgumentException: limit excceed: 67 at com.taobao.tddl.dbsync.binlog.LogBuffer.getLong64(LogBuffer.java:873) at com.taobao.tddl.dbsync.binlog.event.QueryLogEvent.unpackVariables(QueryLogEvent.java:566) ... 6 more ]</description>
		</buginformation>
		<fixedfiles>
			<file>dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/QueryLogEvent.java</file>
		</fixedfiles>
	</bug>
	<bug id="112" opendate="2015-03-30 09:59:58" fixdate="2015-04-10 04:39:58">
		<buginformation>
			<summary>mysql5.6时间毫秒精度支持</summary>
			<description>支持下mysql5.6毫秒精度的解析.  ## 测试case：  CREATE TABLE `t1` (   `id` int(11) NOT NULL AUTO_INCREMENT,   `time0` time DEFAULT NULL,   `time1` time(1) DEFAULT NULL,   `time2` time(2) DEFAULT NULL,   `time3` time(3) DEFAULT NULL,   `time4` time(4) DEFAULT NULL,   `time5` time(5) DEFAULT NULL,   `time6` time(6) DEFAULT NULL,   `timestamp0` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',   `timestamp1` timestamp(1) NOT NULL DEFAULT '0000-00-00 00:00:00.0',   `timestamp2` timestamp(2) NOT NULL DEFAULT '0000-00-00 00:00:00.00',   `timestamp3` timestamp(3) NOT NULL DEFAULT '0000-00-00 00:00:00.000',   `timestamp4` timestamp(4) NOT NULL DEFAULT '0000-00-00 00:00:00.0000',   `timestamp5` timestamp(5) NOT NULL DEFAULT '0000-00-00 00:00:00.00000',   `timestamp6` timestamp(6) NOT NULL DEFAULT '0000-00-00 00:00:00.000000',   `datetime0` datetime DEFAULT NULL,   `datetime1` datetime(1) DEFAULT NULL,   `datetime2` datetime(2) DEFAULT NULL,   `datetime3` datetime(3) DEFAULT NULL,   `datetime4` datetime(4) DEFAULT NULL,   `datetime5` datetime(5) DEFAULT NULL,   `datetime6` datetime(6) DEFAULT NULL,   PRIMARY KEY (`id`) ## ) ENGINE=InnoDB AUTO_INCREMENT=7 DEFAULT CHARSET=utf8  测试数据: 1. insert into t1 values(null,'00:00:00.0','00:00:00.1','00:00:00.02','00:00:00.003','00:00:00.1004','00:00:00.1005','00:00:00.101016','2015-03-30 16:42:39.0','2015-03-30 16:42:39.0','2015-03-30 16:42:39.01','2015-03-30 16:42:39.032','2015-03-30 16:42:39.1023','2015-03-30 16:42:39.10132','2015-03-30 16:42:39.121232','2015-03-30 16:42:39.0','2015-03-30 16:42:39.0','2015-03-30 16:42:39.01','2015-03-30 16:42:39.032','2015-03-30 16:42:39.1023','2015-03-30 16:42:39.10132','2015-03-30 16:42:39.121232'); 2. insert into t1 values(null,'16:42:39.0','16:42:39.1','16:42:39.02','16:42:39.003','16:42:39.1004','16:42:39.1005','16:42:39.101016','2015-03-30 16:42:39.0','2015-03-30 16:42:39.0','2015-03-30 16:42:39.01','2015-03-30 16:42:39.032','2015-03-30 16:42:39.1023','2015-03-30 16:42:39.10132','2015-03-30 16:42:39.121232','2015-03-30 16:42:39.0','2015-03-30 16:42:39.0','2015-03-30 16:42:39.01','2015-03-30 16:42:39.032','2015-03-30 16:42:39.1023','2015-03-30 16:42:39.10132','2015-03-30 16:42:39.121232') 3. insert into t1 values(null,'-16:42:39.0','-16:42:39.1','-16:42:39.02','-16:42:39.003','16:42:39.1004','-16:42:39.1005','-16:42:39.101016','2015-03-30 16:42:39.0','2015-03-30 16:42:39.0','2015-03-30 16:42:39.01','2015-03-30 16:42:39.032','2015-03-30 16:42:39.1023','2015-03-30 16:42:39.10132','2015-03-30 16:42:39.121232','2015-03-30 16:42:39.0','2015-03-30 16:42:39.0','2015-03-30 16:42:39.01','2015-03-30 16:42:39.032','2015-03-30 16:42:39.1023','2015-03-30 16:42:39.10132','2015-03-30 16:42:39.121232'); </description>
		</buginformation>
		<fixedfiles>
			<file>dbsync/src/test/java/com/taobao/tddl/dbsync/binlog/DirectLogFetcherTest.java</file>
			<file>dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java</file>
		</fixedfiles>
	</bug>
	<bug id="105" opendate="2014-12-17 14:06:00" fixdate="2014-12-17 14:12:46">
		<buginformation>
			<summary>mysql5.6开启checksum后,基于时间查找位点会找到错误的位置</summary>
			<description>问题描述：在mysql5.6之后版本中，因为引入了checksum机制，会在正常event的末尾加上4字节的checksum，而在做基于时间查找的seek方法中，decoder解析未识别FORMAT格式的event事件，导致没有判断出开启了checksum，所以导致正常的事务头的BEGIN，多了个4个字节，而判断是否为事务头采用了严格的字符串匹配，就因为多了4个字节，导致event类型出错。  影响范围：基于mysql5.6 + 开启了checksum + 涉及时间查找(比如主备切换和基于时间戳启动)  修改: 1.  增加format类型解析  ```  LogDecoder decoder = new LogDecoder();         decoder.handle(LogEvent.ROTATE_EVENT);         decoder.handle(LogEvent.FORMAT_DESCRIPTION_EVENT);         decoder.handle(LogEvent.QUERY_EVENT);         decoder.handle(LogEvent.XID_EVENT);         LogContext context = new LogContext();         while (fetcher.fetch()) {             LogEvent event = null;             event = decoder.decode(fetcher, context);              if (event == null) {                 throw new CanalParseException("parse failed");             }              if (!func.sink(event)) {                 break;             }         } ``` 1. 基于时间查找，增加事务begin/end的多重判断. </description>
		</buginformation>
		<fixedfiles>
			<file>parse/src/main/java/com/alibaba/otter/canal/parse/inbound/mysql/MysqlConnection.java</file>
		</fixedfiles>
	</bug>
<bugrepository>
